syntax = "proto3";
option java_package = "com.rye.receiptcards.proto";


enum Zone {
  NONE = 0;
  HAND = 1;
  PLAYED = 2;
  DECK = 3;
  SPECIAL = 4;
}

enum SpecialAction {
  SIDEBOARD = 0;
  CREEP = 1;
  RAND_ARROW = 2;
  LEFT_ARROW = 3;
  FORWARD_ARROW = 4;
  RIGHT_ARROW = 5;
  SHOP_NO_HOLD = 6;
  SHOP_HOLD = 7;
  DRAW_PRIZE_CARD = 8;
}

message UUID {
  fixed64 first64 = 1;
  fixed64 second64 = 2;
}

message Move {
  UUID card_uuid = 1;
  Zone source_zone = 2;
  Zone target_zone = 3;
  uint32 num_down = 4;
  bool from_top = 5;
}

message Req {
  UUID user_uuid = 1;
  enum ReqType {
    SHUFFLE = 0;
    MOVE = 1;
    DRAW = 2;
    DECKS_LIST = 3;
    SELECT_DECK = 4;
    SPECIAL = 5;
    RESUME = 6;
  }
  ReqType req_type = 2;
  Zone draw_to = 3;
  Move move = 4;
  uint32 deck_index = 5;
  SpecialAction special = 6;
}

message Rep {
  bool success = 1;
  DecksInfo decks_info = 2;
  repeated Card new_cards = 3;
  repeated Move moves = 4;
  repeated SpecialAction special_actions = 5;
}

message DecksInfo {
  message DeckInfo {
    string name = 1;
    string game = 2;
  }
  repeated DeckInfo decks = 1;
}

message Card {
  UUID card_uuid = 1;
  string image_uri = 2;
}
